src/snowflake/snowpark/_internal/analyzer/analyzer.py

```diff
@@ -154,83 +154,136 @@ def __init__(self, session: "snowflake.snowpark.session.Session") -> None:
         self.alias_maps_to_use = None
 
     def analyze(
-        self, expr: Union[Expression, NamedExpression], parse_local_name=False
+        self,
+        expr: Union[Expression, NamedExpression],
+        fake_col_name_to_real_col_name: Dict[str, str],
```

Comment:
I understand passing in `fake_col_name_to_real_col_name` is causing a lot of changes but this is how I believe recursion should behave. I don't like the approach of `expr_to_alias` where it's an attribute of the Analyzer object (which is session-scoped) and gets reset for each call to `Analyzer.resolve`, I find it complicated and hard to debug. In addition, `Analyzer.analyze` is sometimes used indepdently of `Analyzer.resolve` (e.g. in SQL simplifier), so maybe the best thing to do here is add an additional parameter called `analysis_context` and stuff all the context dependent information there?

Reply:
How about making fake_col_name_to_real_col_name the last parameter and defaults to be `None`? It's more an incremental way than adding it in the middle so it's less likely to break existing code.

Reply:
Well, all calls need `fake_col_name_to_real_col_name ` so it looks right to be in the middle. Forget it.